File: Listas, Colas y Pilas/Colas/ColasListas.cpp
#include <iostream>
using namespace std;

class Nodo {
public:
    int data;
    Nodo* next;
    Nodo(int d) : data(d), next(nullptr) {}
};

class ColaLista {
private:
    Nodo* front;
    Nodo* rear;
public:
    ColaLista() : front(nullptr), rear(nullptr) {}

    bool isEmpty() {
        return front == nullptr;
    }

    void enqueue(int data) {
        Nodo* nuevo = new Nodo(data);
        if (rear == nullptr) {
            front = rear = nuevo;
            return;
        }
        rear->next = nuevo;
        rear = nuevo;
    }

    int dequeue() {
        if (isEmpty()) {
            cout << "Cola vacía (Underflow)\n";
            return -1;
        }
        int d = front->data;
        Nodo* temp = front;
        front = front->next;
        if (front == nullptr) rear = nullptr;
        delete temp;
        return d;
    }

    int peek() {
        if (isEmpty()) {
            cout << "Cola vacía\n";
            return -1;
        }
        return front->data;
    }
};

int main() {
    ColaLista cola;
    cola.enqueue(10);
    cola.enqueue(20);
    cola.enqueue(30);

    cout << "Elemento frontal: " << cola.peek() << endl;
    cout << "Elimina: " << cola.dequeue() << endl;
    cout << "Nuevo frontal: " << cola.peek() << endl;
}

File: Listas, Colas y Pilas/Colas/ColasListas.py
class Nodo:
    def __init__(self, data):
        self.data = data
        self.next = None

class ColaLista:
    def __init__(self):
        self.front = None
        self.rear = None

    def isEmpty(self):
        return self.front is None

    def enqueue(self, data):
        nuevo = Nodo(data)
        if self.rear is None:
            self.front = nuevo
            self.rear = nuevo
            return
        self.rear.next = nuevo
        self.rear = nuevo

    def dequeue(self):
        if self.isEmpty():
            print("Cola vacía (Underflow)")
            return -1
        data = self.front.data
        self.front = self.front.next
        if self.front is None:
            self.rear = None
        return data

    def peek(self):
        if self.isEmpty():
            print("Cola vacía")
            return -1
        return self.front.data

cola = ColaLista()
cola.enqueue(10)
cola.enqueue(20)
cola.enqueue(30)

print("Elemento frontal:", cola.peek())
print("Elimina:", cola.dequeue())
print("Nuevo frontal:", cola.peek())
